File: 1015 KMP/main.py
from __future__ import print_function


def _get_next(pattern, allow_overflapping=False):
    # type (str) -> List[int]
    next_ = [-1]
    i = 0
    j = -1
    while i < len(pattern) - 1:
        if j == -1 or pattern[i] == pattern[j]:
            i += 1
            j += 1
            if not allow_overflapping and pattern[i] == pattern[j]:
                # skip ahead
                next_.append(next_[j])
            else:
                next_.append(j)
        else:
            j = next_[j]

    return next_


def kmp_count(original, pattern, allow_overflapping=True):
    # type: (str, str) -> int

    i = 0
    j = 0
    count = 0
    next_ = _get_next(pattern, allow_overflapping)
    while i < len(original) and j < len(pattern):
        if j == -1 or original[i] == pattern[j]:
            i += 1
            j += 1
        else:
            j = next_[j]

        if j == len(pattern):
            count += 1
            j = next_[j - 1] + 1

    return count


N = int(raw_input())

for _ in range(N):
    p = raw_input()
    o = raw_input()
    print(kmp_count(o, p))

File: 1015 KMP/main.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

vector<int> getNext(string p)
{
    vector<int> next = vector<int>(p.length());
    next[0] = -1;

    int i = 0;
    int j = -1;
    int m = p.length();

    while (i < m - 1)
    {
        if (j == -1 || p[i] == p[j])
        {
            next[++i] = ++j;
        }
        else
        {
            j = next[j];
        }
    }

    return next;
}

int kmp_count(string o, string p)
{
    int n = o.length();
    int m = p.length();

    int i = 0;
    int j = 0;
    int count = 0;

    vector<int> next = getNext(p);

    while (i < n)
    {
        if (o[i] == p[j])
        {
            if (j == m - 1)
            {
                count += 1;
                j = next[j];
            }
            else
            {
                ++i;
                ++j;
            }
        }
        else
        {
            j = next[j];
        }

        if (j == -1)
        {
            ++i;
            j = 0;
        }
    }

    return count;
}

int main()
{
    int N;
    cin >> N;

    for (int i = 0; i < N; ++i)
    {
        string p, o;
        cin >> p >> o;
        cout << kmp_count(o, p) << endl;
    }

    return 0;
}
